### ISSUE TITLE
Batch Seeding Performance

### ISSUE THREAD

**Issue by bzvikler** [Nozbe/WatermelonDB#410]
I'm currently running into issues while trying to use `batch` to seed my database. I have collections of around `50K` records, which caused me to run into memory issues I assume are mentioned in #308.

Since I'm using `batch` and not `sync` I added my own limit of around `1000` to get around the memory issue. This however has lead to very slow initial seeding performance (taking minutes to populate fully).

#403 mentions using the new flag to achieve the `5K` batch size, but I'm not sure if they've run into the slow seeding performance or not.

Anybody had to deal with this case, or the general case of populating watermelon with many records?

**Comment by vinaynarayankutty** [comment]
I need help with this problem #409.

**Comment by radex** [comment]
> taking minutes to populate fully

whoa, that's crazy. :/ our biggest databases are on the order of 20K records and while seeding performance is not great, it's not minutes.

@bzvikler How willing are you to help out and contribute to 🍉, especially on native side?

You can probably work around your issues by seeding to DB from native side… but the best way to improve this situation for everybody would be to improve 🍉. I have a proof of concept JSI rewrite of the native module that's 2-3x faster, and have many more ideas that should give us a big boost as well… but no time to finish this

**Comment by bzvikler** [comment]
Thanks for the response @radex, I'm potentially into contributing to 🍉but will have to get back to you on that, as I'm working for a team currently and we have to evaluate the time it would take to implement changes for the community. 

In the meantime, seeing as your performance for seeding is not in the minutes, it may be the case that I am simply overlooking something in my code that could lead to faster performance, would you be okay with looking at a small snippet of the logic to see if there's something glaring to you?

`const asyncPopulateDatabase = async (db: any): Promise<null> => {
  return new Promise(async (resolve, reject) => {
    try {
      await db.action(async () => {
        await db.unsafeResetDatabase();

        const modelBuilderFactory = new ModelBuilderFactory(db);
        const allPendingRecords = rawConsolidatedData
          .map(({ tableName, records }) => {
            return records
              .map(modelBuilderFactory.getModelBuilder(tableName).buildModel);
          });

        await Promise.all(allPendingRecords.map(batch(db)));
      });

      resolve();
    } catch (e) {
      console.warn(e.message);

      reject(e.message);
    }
  });
};`

it seems pretty simple but maybe there's some optimization step I should be taking, that I'm missing.  Thanks for the help!

**Comment by radex** [comment]
> It seems pretty simple but maybe there's some optimization step I should be taking, that I'm missing.

your approach seems fairly reasonable. Here's what I'd do:

I'd do `poor man's profiling` (`new Date()` before and after, then compare) to time the parts of the process: creating allPendingRecords, calling db.batch, and then inside batch, time how much inside https://github.com/Nozbe/WatermelonDB/blob/master/src/adapters/sqlite/index.js#L202 the `Native.batch` call takes. For more comparable results I'd disable `await Promise.all(allPendingRecords.map(batch(db)));` and batch one by one instead of simultaneously.

so hopefully we'll see if it's the native SQL call that's the main bottleneck, some other part of 🍉, or something in your code

**Comment by esbenp** [comment]
We are currently dealing with the same issue. We also have somebody with 50k+ rows. I did not really dive into it, but I could tell from the debug statements that the time was increasing per 5k batch, e.g. here are our statements from this particular user run in debug mode on simulator

```
[DB] Executed batch of 5000 operations (first: create on order_lines) in 30662.830000001122ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 35314.71500000043ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 44730.94000000128ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 60730.759999998554ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 81707.13500000056ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 112219.94000000268ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 148928.9000000026ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 191319.28500000143ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 242204.27000000203ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 297262.3550000026ms
[DB] Executed batch of 2737 operations (first: create on order_lines) in 331494.12500000105ms
```

And my first intuition was that it might be the listeners that are fired after the batch native batch call. So I just tried to re-run using the same data but outcommenting all the `changeSet` calls in the database module.

```
[DB] Executed batch of 5000 operations (first: create on order_lines) in 31753.979999997682ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 35858.72499999823ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 44185.01999999717ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 59186.610000000655ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 80664.15499999857ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 111960.16499999678ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 148737.98499999975ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 192694.95000000097ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 243281.67000000027ms
[DB] Executed batch of 5000 operations (first: create on order_lines) in 301447.9949999986ms
[DB] Executed batch of 2737 operations (first: create on order_lines) in 338110.31000000006ms
```

As shown there is not really that much of a difference, so right now I am thinking its somewhere in native land

**Comment by esbenp** [comment]
I have created a minimal demo of the issue using the Watermelon native example: https://github.com/esbenp/WatermelonDB/commit/bb27879ea39c2b92ef0fe344339d4c54ec6fd9b9#diff-2eb51f86ebc256f180c86960223e1038

It produces the following times (using comments as the 50k table)

```
[DB] Executed batch of 5000 operations (first: create on comments) in 6361.049999999523ms
[DB] Executed batch of 5000 operations (first: create on comments) in 9071.055000000342ms
[DB] Executed batch of 5000 operations (first: create on comments) in 13841.349999998783ms
[DB] Executed batch of 5000 operations (first: create on comments) in 20548.084999998537ms
[DB] Executed batch of 5000 operations (first: create on comments) in 29547.0150000001ms
[DB] Executed batch of 5000 operations (first: create on comments) in 41810.6050000024ms
[DB] Executed batch of 5000 operations (first: create on comments) in 57741.054999998596ms
[DB] Executed batch of 5000 operations (first: create on comments) in 78426.3749999991ms
[DB] Executed batch of 5000 operations (first: create on comments) in 104484.9549999999ms
[DB] Executed batch of 5000 operations (first: create on comments) in 134143.57999999993ms
```

Although the results are not optimal it seems a lot of the time can be found in userland, e.g. in our app the 10th batch had execution time of 300,000 ms whereas the example had 135,000 ms. I tried getting closer to our app example by adding more data per comment ala

```
const body = 'djkadjaskjdaskld jaskld ajskdlasj dlasjdkasj dasldjaslj dksajldask dajslkd jasjdas lkdjasl dsalkdja';

for(i = 0; i < 1000; i++) {
  const postId = uuid()

  pullDataResponse.posts.created.push({
    id: postId,
    blog_id: blogId,
    title: `Post ${i}`,
    subtitle: 'Some subtitle',
    body: 'Some body'
  })

  for(k = 0; k < 50; k++) {
    pullDataResponse.comments.created.push({
      id: uuid(),
      post_id: postId,
      body: body,
      body_1: body,
      body_2: body,
      body_3: body,
      body_4: body,
      body_5: body,
      is_nasty: false
    })
  }
}
```

and the results were

```
[DB] Executed batch of 5000 operations (first: create on comments) in 10681.900000003225ms
[DB] Executed batch of 5000 operations (first: create on comments) in 13441.549999999552ms
[DB] Executed batch of 5000 operations (first: create on comments) in 18170.234999997774ms
[DB] Executed batch of 5000 operations (first: create on comments) in 25350.1449999967ms
[DB] Executed batch of 5000 operations (first: create on comments) in 35690.84499999735ms
[DB] Executed batch of 5000 operations (first: create on comments) in 53053.930000001856ms
[DB] Executed batch of 5000 operations (first: create on comments) in 73599.73499999978ms
[DB] Executed batch of 5000 operations (first: create on comments) in 97715.5700000003ms
[DB] Executed batch of 5000 operations (first: create on comments) in 127616.46499999915ms
[DB] Executed batch of 5000 operations (first: create on comments) in 159969.71499999927ms
```

so slower, but not significantly so. Here is the same one without `post_id` index

```
[DB] Executed batch of 5000 operations (first: create on comments) in 11707.044999999198ms
[DB] Executed batch of 5000 operations (first: create on comments) in 21472.84999999829ms
[DB] Executed batch of 5000 operations (first: create on comments) in 26354.95499999888ms
[DB] Executed batch of 5000 operations (first: create on comments) in 35505.799999999ms
[DB] Executed batch of 5000 operations (first: create on comments) in 45958.050000001094ms
[DB] Executed batch of 5000 operations (first: create on comments) in 60438.69999999879ms
[DB] Executed batch of 5000 operations (first: create on comments) in 79429.33500000072ms
[DB] Executed batch of 5000 operations (first: create on comments) in 103056.91500000103ms
[DB] Executed batch of 5000 operations (first: create on comments) in 131442.4199999994ms
```

**Comment by radex** [comment]
@esbenp have you tried running the same data additions directly - i.e. run batches of 5k directly, without going through sync? I mean to see if it's only going to be a little faster (the problem is in bridging or somewhere deep in watermelon) or a lot faster (synchronization code itself is a bottleneck)

**Comment by esbenp** [comment]
I did a small test yesterday that tested how much time was spent in (1) get records to apply, (2) apply remote changes to collection and (3) db.batch and it was by far db.batch. I didnt have time to dive in further yet, but I will try to do the same in db.batch to figure out what parts of db.batch are the slow ones

**Comment by radex** [comment]
Please do! See my comment here: https://github.com/Nozbe/WatermelonDB/issues/410#issuecomment-512783760 -- the important part is measuring Native.batch in SQLiteAdapter to see how much time is actually spent in bridge / native code, then we can continue researching this

**Comment by radex** [comment]
btw. I pushed a branch with an experimental rewrite of the native module that sped up batches by ~2x, but will most likely need someone's help to finish this as I don't have too much time on my hands: https://github.com/Nozbe/WatermelonDB/pull/426/files

**Comment by esbenp** [comment]
I don't know much about JSI but is that not part of the RN architectural rewrite? And would that not mean we cannot use it in Watermelon until RN itself uses it (which could be whenever)?

**Comment by radex** [comment]
@esbenp no, JSI already works :)

**Comment by esbenp** [comment]
Alright, let me rephrase: if somebody finished the JSI PR would you switch Watermelon to run on it already? I was not aware it is ready for production-use but I am all for it if it is :-)

**Comment by radex** [comment]
> Alright, let me rephrase: if somebody finished the JSI PR would you switch Watermelon to run on it already? I was not aware it is ready for production-use but I am all for it if it is :-)

Yes, sure! I think it works since 0.58 or 0.59. TurboModules and Fabric, which are based on JSI are not ready yet, but RN already uses JSI, just under the hood -- so it is production.

Or maybe I'd have it as an option for a while. Point is -- I'd run it myself in production.

**Comment by esbenp** [comment]
Very interesting! Maybe I will take a look at it. What do you think it is currently missing for a more finished state?

Btw here is a quick test of where time i spent

```
adapter batch: 717689
change set operations: 62
create operations: 648
get all records to apply: 351
prepare batch: 322
```

so its definitely some sort of native bottleneck

**Comment by radex** [comment]
Ouch! I'm actually surprised how bad this looks. I was sure JS was going to be a major bottleneck as well. I think in that case we should accelerate JSI adoption.

one more thing: was this tested on iOS or Android?

**Comment by esbenp** [comment]
This is very quickly done so I might have done it incorrect, so I will try to do more thorough testing soon.

This is iOS, we don't support Android at all, so I wont get around to try it

**Comment by bzvikler** [comment]
@radex, I pulled down your `rsi-poc` branch. Can you let me know how to get it running, to see the performance difference on my own machine? I don't see the test functions in the `.js` file exported anywhere to be called.

Also, you had mentioned being able to seed the DB on the native side earlier on in this thread, any tips / somewhere you can point me to look into it?

**Comment by radex** [comment]
@bzvikler re: `jsi-poc` -- there's some missing stuff there. I'll try to make it actually compile and be runnable sometime in the next two weeks -- then hopefully someone can take over and polish it until it's done

re: seeding from native code, you need to be familiar with a little bit of Swift/Kotlin to do this. For Swift, see:

- https://github.com/Nozbe/WatermelonDB/blob/master/native/ios/WatermelonDB/DatabaseDriver.swift#L88 - you call this method with a proper list of creates (follow the type trail or take a look at native tests and you'll see what you need to put there). To get a DatabaseDriver, you need to use https://github.com/Nozbe/WatermelonDB/blob/master/native/ios/WatermelonDB/DatabaseBridge.swift -- `connections` has a list of connections, indexed by this tag: https://github.com/Nozbe/WatermelonDB/blob/master/src/adapters/sqlite/index.js#L79 (you need to get ID of database from JS, pass it to native, and use native module to get the Driver). Good luck! If you succeed, it would be very useful if you wrote it up in docs → pro tips for future generations :)

**Comment by bzvikler** [comment]
@radex, wouldn't using the `batch` functionality within swift still have the performance issues, as @esbenp noted that the bottleneck is within the `Native.batch` call?

**Comment by radex** [comment]
> @radex, wouldn't using the batch functionality within swift still have the performance issues, as @esbenp noted that the bottleneck is within the Native.batch call?

A little bit, but only on very large datasets — most of the bottleneck is not SQLite but RN bridge

> Do you think that would work with how watermelon needs its own models and such? I'm optimistically thinking it would be possible to load in a .db file as long as the structure adheres to what watermelon would create via all the Model objects?

Absolutely! This will work without issues.

> If the above would work, then we could look into being able to pass a path to db file, instead of just a db name as is currently the case.

I bring good news: this already works! Just pass a path into `dbName`. I would highly encourage you to contribute a note about this in the docs → pro tips. and a rename in JS code to dbPath would also be nice.

**Comment by radex** [comment]
@bzvikler let me know if version 0.15 improves things. I think you should see a big difference. Next release will have an even greater difference

**Comment by pabloprazeres** [comment]
> I bring good news: this already works! Just pass a path into dbName. I would highly encourage you to contribute a note about this in the docs → pro tips. and a rename in JS code to dbPath would also be nice.

@radex is this available for android? 

I'm facing problems to seed the initial database with batch commands, i have a table with 100k records and some others with 10 to 20k, i think i tried everything i could to make this work, on a weak android 5.1 device with 2gb ram + dual core it takes more than a minute for each 10k record batch command, eventually the app crashes...


Tried passing a path into dbName but got:

`[DB] Uh-oh. Database failed to load, we're in big trouble Error: unknown error (code 14 SQLITE_CANTOPEN): Could not open database`

my db file is in 'android/app/src/main/assets/www/test.db', so in dbName for SQLiteAdapter i got '/android/app/src/main/assets/www/test.db'

is something im doing wrong?

**Comment by radex** [comment]
>  it takes more than a minute for each 10k record batch command, eventually the app crashes...

that's outrageously slow. have you tried using the latest Watermelon prerelease?

**Comment by pabloprazeres** [comment]
> that's outrageously slow. have you tried using the latest Watermelon prerelease?
Yes, already tried 0.14.0, 0.15.0 and 0.16.0-9, didn't noticed any difference

**Comment by BruceSuperProgramer** [comment]
Is watermelon DB is multi-thread? If yes, anyone knows why react-native UI is still laggy(unresponsive to click event for a few seconds) when there are 10k records inserted with batch action.

**Comment by radex** [comment]
@BruceSuperProgramer yes and no. when inserting records, most time is spent passing data between JS and native, so there isn't a lot of opportunity for multithreading. Try out the latest v23 pre-release with jsi mode turned on - you should see batch being >2x faster. It also rearchitects batching such that moving much of the processing to a separate thread (in a future release) should be possible and practical to do.
